ktor-server/ktor-server-plugins/ktor-server-forwarded-header/jvmAndNix/src/io/ktor/server/plugins/ForwardHeaderSupport.kt

```diff
@@ -104,60 +77,151 @@ public object XForwardedHeaderSupport :
         /**
          * Port X-header names. Default is `X-Forwarded-Port`
          */
-        @OptIn(InternalAPI::class)
-        @PublicAPICandidate("2.0.0")
-        internal val portHeaders: ArrayList<String> = arrayListOf("X-Forwarded-Port")
+        public val portHeaders: ArrayList<String> = arrayListOf("X-Forwarded-Port")
+
+        internal var xForwardedHeadersHandler: (MutableOriginConnectionPoint, XForwardedHeaderValues) -> Unit =
+            { _, _ -> }
+
+        init {
+            useFirstValue()
+        }
+
+        /**
+         * Custom logic to extract value from X-Forward-* headers when multiple values are present.
+         * Users need to modify [MutableOriginConnectionPoint] based on headers from [XForwardedHeaderValues]
+         */
+        public fun extractValue(block: (MutableOriginConnectionPoint, XForwardedHeaderValues) -> Unit) {
+            xForwardedHeadersHandler = block
+        }
+
+        /**
+         * Takes first value from X-Forward-* headers when multiple values are present
+         */
+        public fun useFirstValue() {
+            extractValue { connectionPoint, headers ->
+                setValues(connectionPoint, headers) { it.firstOrNull()?.trim() }
+            }
+        }
+
+        /**
+         * Takes last value from X-Forward-* headers when multiple values are present
+         */
+        public fun useLastValue() {
+            extractValue { connectionPoint, headers ->
+                setValues(connectionPoint, headers) { it.lastOrNull()?.trim() }
+            }
+        }
+
+        /**
+         * Takes [proxiesCount]-before-last value from X-Forward-* headers when multiple values are present
+         */
+        public fun skipLastProxies(proxiesCount: Int) {
+            extractValue { connectionPoint, headers ->
+                setValues(connectionPoint, headers) { values ->
+                    values.getOrElse(values.size - proxiesCount - 1) { values.lastOrNull() }?.trim()
+                }
+            }
+        }
+
+        /**
+         * Takes first from the end value that is not in known [hosts]
+         * from X-Forward-* headers when multiple values are present
```

Comment:
Please rephrase the second part or split it into two sentences. Cannot understand the meaning.